src/arch/ucx/machine.C

```diff
@@ -135,10 +136,125 @@ void UcxRequestInit(void *request)
+
+    } else if (tag & UCX_RMA_TAG_DEREG_AND_ACK) {
+        NcpyOperationInfo *ncpyOpInfo = (NcpyOperationInfo *)(rxBuf);
+        resetNcpyOpInfoPointers(ncpyOpInfo);
+        ncpyOpInfo->freeMe = CMK_FREE_NCPYOPINFO;
+
+        if(CmiMyNode() == CmiNodeOf(ncpyOpInfo->srcPe)) { // source node
+            LrtsDeregisterMem(ncpyOpInfo->srcPtr,
+                              ncpyOpInfo->srcLayerInfo + CmiGetRdmaCommonInfoSize(),
+                              ncpyOpInfo->srcPe,
+                              ncpyOpInfo->srcRegMode);
+
+            ncpyOpInfo->isSrcRegistered = 0; // Set isSrcRegistered to 0 after de-registration
+
+            // Invoke source ack
+            if(ncpyOpInfo->opMode != CMK_BCAST_EM_API) {
+                ncpyOpInfo->opMode = CMK_EM_API_SRC_ACK_INVOKE;
+                CmiInvokeNcpyAck(ncpyOpInfo);
+            }
+
+        } else if(CmiMyNode() == CmiNodeOf(ncpyOpInfo->destPe)) { // destination node
+
+            LrtsDeregisterMem(ncpyOpInfo->destPtr,
+                              ncpyOpInfo->destLayerInfo + CmiGetRdmaCommonInfoSize(),
+                              ncpyOpInfo->destPe,
+                              ncpyOpInfo->destRegMode);
+
+            ncpyOpInfo->isDestRegistered = 0; // Set isDestRegistered to 0 after de-registration
+
+            // Invoke destination ack
+            ncpyOpInfo->opMode = CMK_EM_API_DEST_ACK_INVOKE;
+            CmiInvokeNcpyAck(ncpyOpInfo);
+
+        } else {
+            CmiAbort(" Cannot de-register on a different node than the source or destinaton");
+        }
+    }
+#endif
+
+    if (!(tag & UCX_RMA_TAG_MASK)) {
+        handleOneRecvedMsg(size, (char*)rxBuf);
+    }
+
+    if (tag & UCX_MSG_TAG_DIRECT) {
+        // Preposted RX request will be released in UcxPostRxBuffers
+        ++ucxCtx.numFreeRxReqs;
+    } else {
+        UCX_REQUEST_FREE(request);
+    }
+}
+
+
+static inline UcxRequest* UcxPostRxReq(ucp_tag_t tag, size_t size,
+                                       ucp_tag_message_h msg)
+{
+    void *buf = CmiAlloc(size);
+    UcxRequest *req;
+
+    if (tag == UCX_MSG_TAG_DIRECT) {
+        req = (UcxRequest*)ucp_tag_recv_nb(ucxCtx.worker, buf,
+                                           ucxCtx.eagerSize,
+                                           ucp_dt_make_contig(1), tag,
+                                           UCX_MSG_TAG_MASK,
+                                           UcxRxReqCompleted);
+    } else {
+        CmiEnforce(tag == UCX_MSG_TAG_PROBE);
+        req = (UcxRequest*)ucp_tag_msg_recv_nb(ucxCtx.worker, buf, size,
+                                               ucp_dt_make_contig(1), msg,
+                                               UcxRxReqCompleted);
+    }
+
+    CmiEnforce(!UCS_PTR_IS_ERR(req));
+    UCX_LOG(3, "Posted RX buf %p size %zu, req %p, tag %zu, comp %d\n",
+            req->msgBuf, size, req, tag, req->completed);
+
+    // Request completed immediately
+    if (req->completed) {
```

Comment:
I was thinking that it'll aid readability if we have the same condition for immediate completion. 

Isn't `req->completed` equivalent to `reg->msgBuf == NULL`?

Reply:
Not exactly, because when `ucp_tag_recv_nb` returns a request we have to check whether it is already completed or not. We check `req->completed`, since it is filled in the completion callback. An alternative to using `req->completed` field, would be using `ucp_request_check_status` API.  
```
if (ucp_request_check_status(req) != UCS_INPROGRESS) {
}
```
In this case we will not need `req->completed` field.

Reply:
Okay, understood.